### PR TITLE
feat(connector-sendgrid-email): add sendgrid email connector

### PR DESCRIPTION
**Pull request by darcyYe**
## Summary

Add sendgrid email connector.

## Linear Issue Reference

LOG-1612

## Testing

UTs.
Tested locally and received emails.
image
image

### CODE REVIEW

File: packages/connector-sendgrid-mail/src/types.ts [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+const templateGuard = z.object({
+  usageType: z.string(),
+  type: z.nativeEnum(ContextType),
+  subject: z.string(),
+  content: z.string(), // With variable {{code}}, support HTML
+});
+
+export const sendGridMailConfigGuard = z.object({
+  apiKey: z.string(),
+  fromEmail: z.string(), // TODO: can move RegExps to logto/shared and add it to the config.fromEmail schema
```
Comment (wangsijie) [inline packages/connector-sendgrid-mail/src/types.ts:105]: this TODO can be removed, we don't need to check email address in connector, we will check it before calling connector's methods.

File: packages/connector-sendgrid-mail/src/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import { GetConnectorConfig } from '@logto/connector-types';
+
+import { SendGridMailConnector } from '.';
+import { mockedConfig } from './mock';
+import { ContextType, SendGridMailConfig } from './types';
+// Import { request } from './utils';
```
Comment (demonzoo) [inline packages/connector-sendgrid-mail/src/index.test.ts:6]: Remove this?

File: packages/connector-sendgrid-mail/README.md [hunk 3]
```diff
@@ -0,0 +1,2 @@
+### SendGrid Mail README
+placeholder
```
[the remark this answers is not in the document]
Comment (darcyYe) [reply packages/connector-sendgrid-mail/README.md:2]: [LOG-2495](https://linear.app/silverhand/issue/LOG-2495/connectors-readmemd) is added to track this. More sub-issues for each connector will be attached to this issue.

File: packages/connector-sendgrid-mail/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+import {
+  sendGridMailConfigGuard,
+  SendEmailResponse,
+  SendGridMailConfig,
+  EmailData,
+  Personalization,
+  Content,
+  PublicParameters,
+} from './types';
+// Import { request } from './utils';
```
Comment (demonzoo) [inline packages/connector-sendgrid-mail/src/index.ts:23]: Also here?

Comment (darcyYe) [reply packages/connector-sendgrid-mail/src/index.ts]: Removed.

File: packages/connector-sendgrid-mail/src/index.test.ts (hunk 2 above)
Comment (darcyYe) [reply packages/connector-sendgrid-mail/src/index.test.ts]: Removed.

File: packages/connector-sendgrid-mail/docs/config-template.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
+```json
+{
+  "apiKey": "apikey",
```
[the remark this answers is not in the document]
Comment (darcyYe) [reply packages/connector-sendgrid-mail/docs/config-template.md]: fixed.

File: packages/connector-sendgrid-mail/src/types.ts [hunk 6]
```diff
@@ -0,0 +1,119 @@
+import { Nullable } from '@silverhand/essentials';
+import { z } from 'zod';
+
+/**
+ * @doc https://docs.sendgrid.com/api-reference/mail-send/mail-send#body
+ */
+
+export enum ContextType {
+  'TEXT' = 'text/plain',
+  'HTML' = 'text/html',
```
Comment (demonzoo) [inline packages/connector-sendgrid-mail/src/types.ts:10]: We use pascal case in enum keys

File: packages/connector-sendgrid-mail/src/types.ts [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+export type EmailData = {
+  name?: string;
+  email: string;
+};
+
+export type Personalization = {
+  to: EmailData[];
```
Comment (demonzoo) [inline packages/connector-sendgrid-mail/src/types.ts:19]: Should `to` be `EmailData | EmailData[]` as well?

File: packages/connector-sendgrid-mail/src/types.ts [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+};
+
+export type Content = {
+  type: string; // Should either be 'text/html' or 'text/plain'.
+  value: string;
+};
+
+export type Attachment = {
+  content: string;
+  type: string; // Should either be 'text/html' or 'text/plain'
```
Comment (demonzoo) [inline packages/connector-sendgrid-mail/src/types.ts:38]: Can we use string union as the type? e.g. `type: 'text/html' | 'text/plain'?

File: packages/connector-sendgrid-mail/src/types.ts [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+  filename: string;
+  disposition: string; // 'inline' | 'attachment'
```
Comment (demonzoo) [inline packages/connector-sendgrid-mail/src/types.ts:40]: Same here. Can we use `disposition: 'inline' | 'attachment'`?

File: packages/connector-sendgrid-mail/src/types.ts (hunk 7 above)
Comment (darcyYe) [reply packages/connector-sendgrid-mail/src/types.ts:19]: Good call. It shows EmailData | EmailData[] in the doc, but threw an error when requesting with the EmailData type.

File: packages/connector-sendgrid-mail/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    const config = await this.getConfig(this.metadata.target, this.metadata.platform);
+    await this.validateConfig(config);
+    const { apiKey, fromEmail, fromName, templates } = config;
+    const template = templates.find((template) => template.usageType === type);
+
+    assert(
+      template,
+      new ConnectorError(
+        ConnectorErrorCodes.TemplateNotFound,
+        `Cannot find template for type: ${type}`
```
Comment (IceHe) [inline packages/connector-sendgrid-mail/src/index.ts:55]: nit: `cannot find` → `template not found` is more reasonable. (`cannot` seems that there is no way to find)

File: packages/connector-sendgrid-mail/src/mock.ts [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+  Content,
+  ContextType,
+  EmailData,
+  Personalization,
+  PublicParameters,
+  SendGridMailConfig,
+} from './types';
+
+const toEmailData: EmailData[] = [{ email: 'foo@logto.io' }];
+const fromEmailData: EmailData = { email: 'noreply@logto.test.io', name: 'Logto Test' };
```
Comment (IceHe) [inline packages/connector-sendgrid-mail/src/mock.ts:11]: maybe better name them `sender` and `receiver`

File: packages/connector-sendgrid-mail/src/types.ts [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+/**
+ * @doc https://docs.sendgrid.com/api-reference/mail-send/mail-send#body
+ */
+
+export enum ContextType {
+  'TEXT' = 'text/plain',
+  'HTML' = 'text/html',
+}
+
+export type EmailData = {
```
Comment (IceHe) [inline packages/connector-sendgrid-mail/src/types.ts:13]: maybe just name it `Email`?

File: packages/connector-sendgrid-mail/src/index.ts (hunk 10 above)
Comment (darcyYe) [reply packages/connector-sendgrid-mail/src/index.ts]: fixed.

File: packages/connector-sendgrid-mail/src/mock.ts (hunk 11 above)
Comment (darcyYe) [reply packages/connector-sendgrid-mail/src/mock.ts]: fixed.

File: packages/connector-sendgrid-mail/src/types.ts (hunk 12 above)
Comment (darcyYe) [reply packages/connector-sendgrid-mail/src/types.ts:13]: Used EmailData to align with provider's docs for better understanding.

File: packages/connector-sendgrid-mail/src/types.ts (hunk 7 above)
Comment (demonzoo) [reply packages/connector-sendgrid-mail/src/types.ts:19]: [facepalm]..

File: packages/connector-sendgrid-mail/docs/config-template.md [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+    },
+    {
+      "usageType": "Register",
+      "type": "ContextType.TEXT",
+      "subject": "Logto Register Template",
+      "content": "This is for registering purposes only. Your passcode is {{code}}.",
+    },
+    {
+      "usageType": "Test",
+      "type": "ContextType.TEXT",
```
Comment (demonzoo) [inline packages/connector-sendgrid-mail/docs/config-template.md:20]: Update enum usage to `ContextType.Text`

File: packages/connector-sendgrid-mail/package.json [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+        "dev": "rm -rf lib/ && tsc-watch -p tsconfig.build.json --preserveWatchOutput --onSuccess \"node ./lib/index.js\"",
+        "test": "jest",
+        "test:coverage": "jest --coverage --silent",
+        "prepack": "pnpm build"
+    },
+    "dependencies": {
+        "@logto/connector-types": "^0.1.0",
+        "@logto/shared": "^0.1.0",
+        "@silverhand/essentials": "^1.1.6",
+        "@silverhand/jest-config": "^0.13.0",
```
Comment (demonzoo) [inline packages/connector-sendgrid-mail/package.json:29]: we can upgrade to 0.14.0 now

File: packages/connector-sendgrid-mail/package.json [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+        "got": "^11.8.2",
+        "zod": "^3.14.3"
+    },
+    "devDependencies": {
+        "@jest/types": "^27.5.1",
+        "@silverhand/eslint-config": "^0.10.2",
+        "@silverhand/ts-config": "^0.11.0",
```
Comment (demonzoo) [inline packages/connector-sendgrid-mail/package.json:36]: 0.14.0

File: packages/connector-sendgrid-mail/src/types.ts [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+  subject?: string;
+  headers?: Record<string, string>;
+  substitutions?: Record<string, string>;
+  dynamic_template_data?: Record<string, any>;
+  custom_args?: Record<string, string>;
+  sendAt?: number;
+};
+
+export type Content = {
+  type: 'text/html' | 'text/plain';
```
Comment (demonzoo) [inline packages/connector-sendgrid-mail/src/types.ts:32]: We can convert `ContextType` enum to a union type, without having to define it again.
```typescript
type: `${ContextType}`
```

File: packages/connector-sendgrid-mail/src/types.ts [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+  value: string;
+};
+
+export type Attachment = {
+  content: string;
+  type: 'text/html' | 'text/plain';
```
Comment (demonzoo) [inline packages/connector-sendgrid-mail/src/types.ts:38]: same here, convert `ContextType` enum to union type

**Review by demonzoo — APPROVED** [review]
Overall looks good, please check my latest comments, thx

**Review by wangsijie — APPROVED** [review]

**Pull request merged**